OPC-621 Paella embed iFrame driver API

karendolan opened the pull request:
This is a companion pull for OPC-621 in bitbucket. It is to support the immersive player feature

karendolan commented:
NOTE: this pull will remains outside of production until Immersive Player development gets approval.

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+/*
+ * DCE IFrameApiHelper
```
Missing author

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.timedCommentsPlugin/timedCommentsHeatmap.js:
```diff
@@ -102,7 +102,10 @@ paella.addPlugin(function () {
+          // OPC-621 Force annots off with special URL param, social=off
+          // for Paella iFrame API embed
+          const socialArg = base.parameters.get('social');
+          onSuccess(! paella.player.isLiveStream() && socialArg !== 'off');
```
No space after `!`

gabeabrams reviewed (changes requested):
I finally started my review! xD Thanks for your patience

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+      })
+    }
+
+    getName() {
```
Yes, @gabeabrams look at the changed files, the JSDoc is there

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    getName() {
+      return "edu.harvard.dce.paella.iFrameApiAppraiser";
```
done

karendolan commented on config/config.json:
```diff
@@ -392,7 +392,9 @@
       },
       "//*****  Keyboard plugins": "",
-      "es.upv.paella.defaultKeysPlugin": {"enabled": true }
+      "es.upv.paella.defaultKeysPlugin": {"enabled": true },
```
closing as won't resolve for now to be consistent with UPV until next upgrade.

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+      return "edu.harvard.dce.paella.iFrameApiAppraiser";
+    }
+
+    checkEnabled (onSuccess) {
```
added JSDocs

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    getEvents() {
+      return[
```
done

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+        default:
+          base.log.debug('IFrameApiAppraiser: Unsupported event ', eventType);
+      }
+      thisClass.lastEvent = eventType;
```
That was an artifact, fixed to just be this

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+      );
+    }
+
+    getEvents() {
```
resolved

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+ * IFrame embed API plugin used in conjunction with iFrameEmbedApi.js
+ * that is imported into parent window.
+ */
+paella.addPlugin(function () {
```
When we upgrade Paella player all that can be addressed.

karendolan commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,85 @@
+    checkEnabled (onSuccess) {
+      // This iFrame prefix is used by the iFrameEmbedApi
+      // When constructing the embeded player iFrame
+      const iFramePrefix = 'DCE-iframe-API';
```
I'm deferring this suggestion for now because this constant is only used within the checkEnabled context of this class.

karendolan commented:
Hi @gabeabrams I know this is really old, I just noticed that this pull hasn't been officially approved. I believe I addressed all your comments from your review last fall.

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /**
+     * @inheritdoc
```
Put tags below text:

```
     /**
      * Get the name of the Paella plugin
      * @inheritdoc
      * @return the name of this plugin
      */
```

gabeabrams replied:
Same below?

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+      * @inheritdoc
+      * Get the name of the Paella plugin
+      * @param {requestCallback} onSuccess - The callback that handles response
+      */
```
Wrong indentation

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+    /**
+      * @inheritdoc
+      * Get the events that this Paella event driven plugin uses
+      * @return an array of events
```
Indentation

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+      * called when an event fires.
+      * @param {string} eventType - the event that was fired
+      * @param {object} [params] - a set of params associated to the event
+      */
```
Indentation

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+        case paella.events.seeked:
+        case paella.events.error:
+        case paella.events.videoReady:
+        case paella.events.loadComplete:
```
Does this really work as intended without `break;` statements?

Perhaps switch this to a set of `ifs`?

karendolan replied:
The switch seemed cleaner than 3 if-thens.
The list of cases are just the fall-through of cases that all fall to the this.sendMessageToEmbedApi(eventType);
Anything not in that list is not supported to send up to the container.

gabeabrams replied:
No I mean that this logic might not work as intended without `break;` statements.

For example, this switch will call this line: `paella.log.debug('IFrameApiAppraiser: Unsupported event ', eventType);` for anything that's not `paella.events.timeupdate` or `paella.events.ended`

karendolan replied:
Line 87 has the break for all the cases above it that don't have a break. They all fall through to the paella.events.ended.  Then hit that break at line 87. That's the beauty of a switch.

karendolan replied:
Switch is cool to do stuff like :
case: is formal dinner
   dress nice
case: is dinner
  eat a light lunch to save room for dinner
  break;
.... 
So the formal dinner gets both dress nice and eat a light lunch.

gabeabrams replied:
Is it intended behavior for `paella.events.videoReady` to print "Unsupported event"?

karendolan replied:
Here's a case that shows 2 case catches getting the same action https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch

gabeabrams replied:
(just an example)

karendolan replied:
> 



> Is it intended behavior for `paella.events.videoReady` to print "Unsupported event"?

If it was the `this.sendMessageToEmbedApi(eventType);` would be moved below the ` this.sendMessageToEmbedApi(eventType);` which is what it's hitting now. Or remove from the CASE so it automatically goes to the default catch.

gabeabrams replied:
If you want `paella.events.play`, `paella.events.pause`, `paella.events.seeked`, `paella.events.error`, `paella.events.videoReady`, and `paella.events.loadComplete` to call `paella.log.debug` and log "Unsupported event" then we're all good and I understand.

karendolan replied:
The break on 87 prevents `paella.events.videoReady` from getting to the debug line.

gabeabrams replied:
Oh wait. I'm not reading this properly. >.<

gabeabrams replied:
Okay I get it. Wow sorry that took me so long :P
I still don't like it haha >.< but I understand it now. You can resolve this convo.

gabeabrams replied:
If I wrote this, I'd probably use something like:

```
// Make sure the event is supported
if (!(eventType in Object.values(paella.events))) {
  paella.log.debug('IFrameApiAppraiser: Unsupported event ', eventType);
}

// Send the notification
if (eventType === paella.events.timeupdate) {
  this.sendMessageToEmbedApi(eventType, params.currentTime);
} else {
  this.sendMessageToEmbedApi(eventType);
}
```

but that's just me.

Again, I get this now. Sorry!

karendolan replied:
I get that. But, there are a lot of events in paella events that are not useful or possibly even outdated and unused. Enumerating the events to pass to the wrapper  gives this plugin control to consolidate the events and protect if the naming changes on the event inside the player. 
There are pros & cons. But as an API, it seems safer to be explicit about what is being be passed to the wrapper.

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+      const newMessage = {
+        sender: iFrameName,  // required
+        name: eventType, // required
+        value,  // Optional
```
Too many spaces before `//`

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+    sendMessageToEmbedApi(eventType, value) {
+      const iFrameName = window.name;
+      const newMessage = {
+        sender: iFrameName,  // required
```
Too many spaces before `//`

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -0,0 +1,112 @@
+      // The plugin sends the above event types to all containers
+      // The Host source of the embedAPI is unknown to the player
+      // Player status information is low risk
+      window.parent.postMessage(newMessage, '*');
```
What's the `*` mean?

karendolan replied:
The * is the target window. It can be any site that wants to wrap the player. There are only approved events being sent the wrapper which don't contain anything secure so it's ok for malicious wrappers to intercept start, stop, pause event notifications.

gabeabrams replied:
Thank you :)

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.timedCommentsPlugin/timedCommentsHeatmap.js:
```diff
@@ -16,7 +16,7 @@ paella.addPlugin(function () {
     }
-    
+
     getAlignment() {
```
JSDoc? But this is not part of your PR

karendolan replied:
That would be a whole other can of worms to start adding JSDocs here. These plugins confirm with Paella 6 that doesn't have JDocs. The Paella 7 will require all these plugins to be revisited anyway.

gabeabrams replied:
Okay

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.timedCommentsPlugin/timedCommentsHeatmap.js:
```diff
@@ -164,16 +167,16 @@ paella.addPlugin(function () {
-      
-      
+
+
```
Just one empty line

gabeabrams reviewed (commented):
Looks good, just some little things that aren't urgent.

karendolan commented:
> Looks good, just some little things that aren't urgent.

@gabeabrams Please let me know if you have any other concerns about this pull. Keeping the locked down API contract will save us, or those who go after us, a lot of stress.

gabeabrams commented on vendor/plugins/edu.harvard.dce.paella.iFrameApiAppraiser/iFrameApiAppraiser.js:
```diff
@@ -103,7 +103,7 @@ paella.addPlugin(function () {
       if (this.eventsToAppraiseWithTime.contains(eventType)) {
         this.sendMessageToEmbedApi(eventType, params.currentTime);
-      } else if (this.eventsToAppraiseWithoutTime) {
+      } else if (this.eventsToAppraiseWithoutTime.contains(eventType)) {
```
If this isn't being cross-compiled, in older browsers, it's best to use `indexOf(...) >= 0` instead of `contains`.
But that's probably not an issue because we require modern browsers.

karendolan replied:
I messed up on that.  I fixed by changing the contains to includes.

gabeabrams reviewed (approved):
Looks awesome!

karendolan commented on tests/iFrameApiAppraiser-tests.js:
```diff
@@ -0,0 +1,210 @@
+var _ = require('lodash');
+var url = require('url');
+var reload = require('require-reload')(require);
+var Promise = require('bluebird');
```
whoah, I don't need these. A cut & paste artifact
